src/scanner.c

```diff
@@ -207,245 +357,291 @@ typedef struct {
+  return is_eof() ? 0 : env->lexer->get_column(env->lexer);
 }
 
 /**
- * Instruct the lexer that the current position is the end of the potentially detected symbol, causing the next run to
- * be started after this character in the success case.
- *
- * This is useful if the validity of the detected symbol depends on what follows, e.g. in the case of a layout end
- * before a `where` token.
+ * Don't add zeroes to the lookahead buffer when hitting EOF – it causes `no_lookahead` to report false negatives.
  */
+static void advance() {
+  if (not_eof()) {
+    VEC_PUSH(state->lookahead, PEEK);
+    env->lexer->advance(env->lexer, false);
+  }
+}
+
+static bool set_result_symbol(Symbol result) {
+  if (result != FAIL) {
+    env->lexer->result_symbol = (TSSymbol) result;
+    return true;
+  }
+  else return false;
+}
 
-// Only use string literals we actually need
-#ifdef DEBUG
-static void MARK(char *marked_by, bool needs_free, State *state) {
-  state->marked = column(state);
-  if (state->needs_free) free(state->marked_by);
-  state->marked_by = marked_by;
-  state->needs_free = needs_free;
-  state->lexer->mark_end(state->lexer);
+#if DEBUG
+
+static void mark_debug(const char *restrict marked_by) {
+  dbg("mark: %s\n", marked_by);
+  env->debug.marked = (int) column();
+  env->debug.marked_line = 0;
+  env->debug.marked_by = marked_by;
+  env->lexer->mark_end(env->lexer);
 }
-#else
-#define MARK(s, nf, state) state->lexer->mark_end(state->lexer);
+
+static void append_parse_buffer();
+
+static void advance_debug() {
+  append_parse_buffer();
+  advance();
+}
+
+static void skip_debug() {
+  append_parse_buffer();
+  env->lexer->advance(env->lexer, true);
+}
+
 #endif
 
+/**
+ * `inline` has a noticeable impact, reaching parity with a macro.
+ */
+static inline bool valid(Symbol s) { return env->symbols[s]; }
+
+// --------------------------------------------------------------------------------------------------------
+// Symbol constructors
+// --------------------------------------------------------------------------------------------------------
+
+static Symbol finish(Symbol s, const char *restrict desc) {
+  // Suppress unused param warning
+  (void) desc;
+  dbg("finish: %s\n", desc);
+  return s;
+}
+
+static Symbol finish_if_valid(Symbol s, const char *restrict desc) {
+  if (valid(s)) return finish(s, desc);
+  else return FAIL;
+}
+
+static Symbol finish_marked(Symbol s, const char *restrict desc) {
+  (void) desc;
+  MARK(desc);
+  return s;
+}
+
+static Symbol update_state(const char *restrict desc) {
+  return finish(UPDATE, desc);
+}
 
 // --------------------------------------------------------------------------------------------------------
-// Condition
+// Lookahead
 // --------------------------------------------------------------------------------------------------------
 
+static void advance_over(uint32_t n) {
+  for (uint32_t i = state->lookahead->len; i <= state->lookahead->offset + n; i++) S_ADVANCE;
```

Comment:
I'm not sure why we're advancing with relation to the peek position, and I'm not convinced this for look is right anyway. Shouldn't the length of the lookahead be a limit?

Reply:
The docs now explain it, but for the record:

`lookahead` is the range that's _already_ been advanced over, so this starts at its end.
`offset` is the reference point for indexing `lookahead` that can be reset at any point for convenience, so the target index is `offset + n`.
If that is smaller than `lookahead`'s current size, the target index has already been consumed and the loop immediately returns.

Reply:
Okay, so there's an internal state, and a tree-sitter state.

`advance`, now pushes the current token to a the internal-state lookahead, and advances the tree-sitter character.

`peek` peeks into a internal lookahead buffer, and may advance the tree-sitter state to fill that buffer, if necessary.

So when parsing a file from scratch, does that mean that the entire file is pushed character-by-character into the lookahead buffer?

By the looks of it `reset_lookahead` doesn't reset the lookahead, so much as it advances to the end of the lookahead buffer. Maybe there's a better name for this operation?

---

For my own interest, do you think it's possible for the scanner to work with only one character of lookahead?

Reply:
> Okay, so there's an internal state, and a tree-sitter state.
> 
> `advance`, now pushes the current token to a the internal-state lookahead, and advances the tree-sitter character.
> 
> `peek` peeks into a internal lookahead buffer, and may advance the tree-sitter state to fill that buffer, if necessary.
> 
> So when parsing a file from scratch, does that mean that the entire file is pushed character-by-character into the lookahead buffer?

If the question is whether the entire file is stored in that buffer _entirely_ at some point, then no – the buffer is discarded after each run.
Right now the array is reused across runs and only overwritten to achieve that, but like the global pointers this isn't thread safe, so I'll have to revert to the initial approach I used and allocate a new one in each run (unlike the state pointers, this would only be unsafe if multiple nodes _in the same file_ would be processed concurrently, so it might still be feasible if we knew that TS doesn't allow that).

It's probably useful to point out that none of this has any impact on performance.
I've never observed any differences in benchmarks when changing implementation details of the scanner on a "technical" level, only when modifying logic so that the _number_ of times the scanner is called varies.
(On that note, moving stuff from the grammar to the scanner usually results in _significant_ perf benefits 😅 e.g. when I moved pragmas in, HLS parse time reduced from 420ms to 360ms)

I've also looked at memory usage but the profiling capabilities are limited with simple tools.
I only measured RSS, which varied between 19MB and 21MB for `ghc/compiler` before and after rewrite.

> By the looks of it `reset_lookahead` doesn't reset the lookahead, so much as it advances the internal state to the end of the lookahead buffer. Maybe there's a better name for this operation?

I don't have a clear idea of what "resetting lookahead" canonically means, so if you feel confident about an alternative I'm happy to change it 🙂
My interpretation is that `peek0()` is the analogue of `env->lexer->lookahead`, and resetting the offset makes the former return the same as the latter.

> For my own interest, do you think it's possible for the scanner to work with only one character of lookahead?

I think it's possible that some things would need to be spread over two runs like newline lookahead, but you could probably implement it in a convoluted way.
The main motivation is to be able to write the individual fragments independently though, without having to pay attention to their order.